src/python/pants/backend/project_info/depgraph.py

```diff
@@ -0,0 +1,187 @@
+
+    def get_dependencies(self, v: Vertex) -> tuple[Vertex, ...]:
+        return tuple(self.vertices_by_id[edge.dep_id] for edge in self.edges_by_src.get(v.id, []))
+
+    def get_dependees(self, v: Vertex) -> tuple[Vertex, ...]:
+        return tuple(self.vertices_by_id[edge.src_id] for edge in self.edges_by_dep.get(v.id, []))
+
+    def to_json(self) -> str:
+        return json.dumps(self, cls=DependencyGraphJSONEncoder, indent=2)
+
+
+class DependencyGraphJSONEncoder(JSONEncoder):
+    def default(self, o):
+        if isinstance(o, DependencyGraph):
+            return {"vertices": o.vertices, "edges": o.edges}
+        elif isinstance(o, Vertex):
+            return {"id": o.id, "data": o.data}
+        elif isinstance(o, Edge):
+            return {"src_id": o.src_id, "dep_id": o.dep_id, "data": o.data}
+        elif isinstance(o, FrozenDict):
+            return dict(o)  # Unfortunately, FrozenDict is not JSON serializable.
+        return super().default(o)
+
+
+@dataclass(frozen=True)
+class DependencyGraphRequest:
+    addresses: Addresses
+    transitive: bool
+
+
+@rule
+async def generate_graph(request: DependencyGraphRequest) -> DependencyGraph:
+    def target_sorter(tgts_to_sort: Iterable[Target]) -> list[Target]:
+        return sorted(tgts_to_sort, key=lambda t: t.address)
+
+    if request.transitive:
+        transitive_targets = await Get(
+            TransitiveTargets,
+            TransitiveTargetsRequest(request.addresses, include_special_cased_deps=True),
+        )
+        targets = set(transitive_targets.roots) | set(transitive_targets.dependencies)
+    else:
+        unexpanded_targets = await Get(UnexpandedTargets, Addresses, request.addresses)
+        targets = set(unexpanded_targets)
+
+    sorted_targets = target_sorter(targets)
+
+    dependencies_per_target = await MultiGet(
+        Get(
+            Targets,
+            DependenciesRequest(tgt.get(Dependencies), include_special_cased_deps=True),
+        )
+        for tgt in sorted_targets
+    )
+
+    id_seq = 0
+    vertices = []
+    vertices_by_tgt = {}
+    # Make sure we create vertices for the dependencies, even in non-transitive mode.
+    vertex_targets = target_sorter(
+        targets | set(itertools.chain.from_iterable(dependencies_per_target))
+    )
+    for tgt in vertex_targets:
+        tgt_id = str(id_seq)
+        id_seq += 1
+        addr = tgt.address
+        vertex_data: dict = {
+            "address": addr.spec,
+            "type": tgt.alias,
+        }
+        # TODO: Code in project_info shouldn't import from, or know about, Python-specific fields.
+        #  We need a generic way to express "this target represents these 3rdparty dep coordinates".
+        if tgt.has_field(PythonRequirementsField):
+            vertex_data["requirements"] = tuple(
+                str(python_req) for python_req in tgt[PythonRequirementsField].value
+            )
```

Comment:
Would be good to link this to the deprecation of `dependencies --type` so that we remember to remove it.

Reply:
I don't necessarily think we want to remove it. It's pretty useful to have some useful subset data on the vertices (and, eventually, edges), although granted what is deemed "useful" is subjective.  But not forcing people to do crazy pipes and xargs for some of the more apparently common cases makes sense to me.